src/lib/proof-system/zkprogram.ts

```diff
@@ -415,8 +442,21 @@ function ZkProgram<
         programState.reset(methodIntfs[i].methodName);
       }
 
+      let publicOutput;
       let [publicOutputFields, proof] = MlPair.from(result);
-      let publicOutput = fromFieldConsts(publicOutputType, publicOutputFields);
+      if (nonPureInputExists) {
+        let nonPureOutput = programState.getNonPureOutput();
+
+        publicOutput = fromFieldConsts(
+          publicOutputType,
+          publicOutputFields,
+          nonPureOutput
+        );
+
+        programState.reset('__nonPureOutput__');
+      } else {
```

Comment:
nit: we could also simplify this logic here if we simply return `[]` or `undefined` from `getNonPureOutput`